coffeecard/CoffeeCard.Models/Entities/Statistic.cs

```diff
@@ -1,31 +1,73 @@
 ﻿using System;
 using System.ComponentModel.DataAnnotations.Schema;
+using Microsoft.EntityFrameworkCore;
 
 namespace CoffeeCard.Models.Entities
 {
-    /**
-    * This class will be used to optimize the queries to get statistics.
-    * The preset defines the kind of statistic entity and each preset has a corresponding `SwipeCount`, `SwipeRank` and `ExpiryDate`.
-    * This means that each user should have exactly 3 entries in the database after 1 or more swipes.
-    * Each preset should be updated accordingly to the current date and the ExpiryDate. If the ExpiryDate of a entry is exceeded, then the `SwipeCount`, `SwipeRank` and `ExpiryDate` should be reset.
-    **/
-    public enum StatisticPreset { Monthly, Semester, Total }
+    /// <summary>
+    /// This class will be used to optimize the queries to get statistics.
+    /// The preset defines the kind of statistic entity and each preset has a corresponding `SwipeCount`, `SwipeRank` and `ExpiryDate`.
+    /// This means that each user should have exactly 3 entries in the database after 1 or more swipes.
+    ///
+    /// Each preset should be updated accordingly to the current date and the ExpiryDate.
+    /// If the ExpiryDate of a entry is exceeded, then the `SwipeCount`, `SwipeRank` and `ExpiryDate` should be reset.
+    /// </summary>
+    public enum StatisticPreset
+    {
+        /// <summary>
+        /// Statistic preset for a month
+        /// </summary>
+        Monthly, 
+        /// <summary>
+        /// Statistic preset for a semester
+        /// </summary>
+        Semester, 
+        /// <summary>
+        /// All Statistics
+        /// </summary>
+        Total
+    }
 
+    /// <summary>
+    /// Statistic represent a user's swipe of tickets. A Statistic entry is valid within a given preset period unless it is the total statistics.
+    /// </summary>
+    [Index(nameof(Preset), nameof(ExpiryDate), nameof(SwipeCount), nameof(LastSwipe))]
```

Comment:
My main concern with this index, would be whether it is too expensive to maintain for the database, compared to the increase in speed it would provide. 

That is, every time a person makes a swipe we would end up having to rebuild the index. Have you done any testing to see how this performs?

I considered adding one to the preset myself, since this should not change, so there would be less overhead in maintaining it. We could also expand it to include the expiry date, since this should change rarely as well.

Reply:
I don't think that it will be expensive compared to the size of the table. I imagined this index would be useful for the Get Sorted Statistics.

I can't recall the documentation entirely but don't think the entire Index is being rebuilt at every swipe, only the relevant entries. Has this been a sorted index then likely it could be that it would be completely rebuilt.